rocAL/rocAL/include/meta_data/meta_data.h

```diff
@@ -81,70 +90,143 @@ typedef struct
     RotationBatch rotation_batch;
 }JointsDataBatch;
 
+typedef struct MetaDataInfo {
+public:
+    uint img_id = -1;
+    std::string img_name = "";
+    ImgSize img_size = {};
+} MetaDataInfo;
+
 struct MetaData
 {
-    int& get_label() { return _label_id; }
-    BoundingBoxCords& get_bb_cords() { return _bb_cords; }
-    BoundingBoxCords_xcycwh& get_bb_cords_xcycwh() { return _bb_cords_xcycwh; }
-    BoundingBoxLabels& get_bb_labels() { return _bb_label_ids; }
-    void set_bb_labels(BoundingBoxLabels bb_label_ids) {_bb_label_ids = std::move(bb_label_ids); }
-    ImgSize& get_img_size() { return _img_size; }
-    const JointsData& get_joints_data(){ return _joints_data; }
+    virtual void get_labels(std::vector<int>** label_ids) {}
+    virtual void set_labels(Labels label_ids) {}
+    virtual void get_bb_cords(BoundingBoxCords** bb_cords) {}
+    virtual void get_bb_cords_xcycwh(BoundingBoxCords_xcycwh** bb_cords_xcycwh) {}
+    virtual void set_bb_cords_xcycwh(BoundingBoxCords_xcycwh bb_cords_xcycwh) {}
+    virtual void set_bb_cords(BoundingBoxCords bb_cords) {}
+    virtual void get_polygon_count(std::vector<int>** polygon_count) {}
+    virtual void get_vertices_count(std::vector<std::vector<int>>** vertices_count) {}
+    virtual void get_mask_cords(MaskCords** mask_cords) {}
+    virtual void set_mask_cords(MaskCords mask_cords) {}
+    virtual void set_polygon_counts(std::vector<int> polygon_count) {}
+    virtual void set_vertices_counts(std::vector<std::vector<int>> vertices_count) {}
+    virtual void get_joints_data(JointsData** joints_data) {}
+    virtual void set_joints_data(JointsData *joints_data) {}
+    ImgSize& get_img_size() {return _info.img_size; }
+    std::string& get_image_name() { return _info.img_name; }
+    uint& get_image_id() { return _info.img_id; }
+    void set_img_size(ImgSize img_size) { _info.img_size = std::move(img_size); }
+    void set_img_id(uint img_id) { _info.img_id = img_id; }
+    void set_img_name(std::string img_name) { _info.img_name = img_name; }
+    void set_metadata_info(MetaDataInfo info) { _info = std::move(info); }
 protected:
-    BoundingBoxCords _bb_cords = {}; // For bb use
-    BoundingBoxCords_xcycwh _bb_cords_xcycwh = {}; // For bb use
-    BoundingBoxLabels _bb_label_ids = {};// For bb use
-    ImgSize _img_size = {};
-    JointsData _joints_data = {};
-    int _label_id = -1; // For label use only
+    MetaDataInfo _info;
+};
+
+template <typename T>
+void (MetaData::*pGetMetaDataFunction)(T** ptr);
+
+template <typename T>
+T& getMetaDataValues(MetaData &metadata, void(MetaData::*pGetMetaDataFunction)(T** ptr))
```

Comment:
I am not sure what is the use of this function? It is very ambiguous. If it is to call a base class function through derived class, you can use b.a::foo() (where b is derived from a). All base class functions are inherited for derived class as well. So why do we need to call it through funcion pointers? Looks like we are defeating the purpose of c++

Reply:
In the getMetadataValues function. We use the function pointer and invoke the function of the derived class only using the object of the derived class that is passed to this function.

In this API we are not invoking a base class function through derived class, Instead we are passing a function pointer of the required function and an object of the Metadata (this object will be an instance of the derived class)
funcName<return_type>(derived object, funcPtr);

Reply:
we need to discuss how to do this properly using c++ api